drv/lpc55-update-server/src/images.rs

```diff
@@ -23,82 +31,605 @@ extern "C" {
+                exec: __IMAGE_B_BASE.as_ptr() as u32
+                    ..__IMAGE_B_END.as_ptr() as u32,
+            },
+            (RotComponent::Stage0, SlotId::A) => FlashRange {
+                store: __IMAGE_STAGE0_BASE.as_ptr() as u32
                     ..__IMAGE_STAGE0_END.as_ptr() as u32,
-                __IMAGE_STAGE0_BASE.as_ptr() as u32
+                exec: __IMAGE_STAGE0_BASE.as_ptr() as u32
                     ..__IMAGE_STAGE0_END.as_ptr() as u32,
-            ),
-            (RotComponent::Stage0, SlotId::B) => (
-                __IMAGE_STAGE0NEXT_BASE.as_ptr() as u32
+            },
+            (RotComponent::Stage0, SlotId::B) => FlashRange {
+                store: __IMAGE_STAGE0NEXT_BASE.as_ptr() as u32
                     ..__IMAGE_STAGE0NEXT_END.as_ptr() as u32,
-                __IMAGE_STAGE0_BASE.as_ptr() as u32
+                exec: __IMAGE_STAGE0_BASE.as_ptr() as u32
                     ..__IMAGE_STAGE0_END.as_ptr() as u32,
-            ),
+            },
+        }
+    }
+}
+
+#[derive(Clone, PartialEq)]
+pub struct Epoch {
+    value: u32,
+}
+
+/// Convert from the ImageHeader.epoch format
+impl From<u32> for Epoch {
+    fn from(number: u32) -> Self {
+        Epoch { value: number }
+    }
+}
+
+/// Convert from the caboose EPOC format.
+impl From<&[u8]> for Epoch {
+    fn from(chars: &[u8]) -> Self {
+        match chars.iter().try_fold(0u32, |epoch: u32, c| {
+            if let Some(digit) = (*c).checked_sub(b'0') {
+                if digit < 10 {
+                    epoch
+                        .checked_mul(10u32)
+                        .and_then(|product| product.checked_add(digit as u32))
+                } else {
+                    None
+                }
+            } else {
+                None
+            }
+        }) {
+            Some(value) => Epoch { value },
+            None => Epoch::from(0u32),
+        }
+    }
+}
+
+impl Epoch {
+    pub fn cannot_update_to(&self, next: Epoch) -> bool {
+        self.value < next.value
+    }
+}
+
+/// The update_server needs to deal with images that are:
+///   - Entirely in flash
+///   - Header block in RAM with the remainder unavailable
+///   - Header block in RAM with the remainder in flash.
+///   - Header block in RAM but also the entire image in RAM (Stage0)
+/// Calls to methods use image relative addresses.
+/// ImageAccess::Flash{_, span} gives the physical flash addresses
+/// being accessed.
+pub enum ImageAccess<'a> {
+    // Flash driver, flash device range
+    Flash {
+        flash: &'a drv_lpc55_flash::Flash<'a>,
+        span: &'a Range<u32>,
+    },
+    Ram {
+        buffer: &'a [u8],
+        // True if this is a complete image.
+        complete: bool,
+    },
+    // The image is ram[0..] concatenated with flash span[ram.len()..]
+    Hybrid {
+        buffer: &'a [u8],
+        flash: &'a drv_lpc55_flash::Flash<'a>,
+        span: &'a Range<u32>,
+    },
+}
```

Comment:
I think we should break this out into a separate PR to review as well to make sure we're getting the abstraction right

Reply:
Done in POR #1813